### DISCUSSION TITLE
Quad meshes & Ngons

### DISCUSSION THREAD

**Discussion by ant1fact** [asny/tri-mesh#20] [Q&A] [answered]
Hello!

I stumbled upon this repository and I find it very interesting. I am a 3d modeller by trade but I enjoy programming on the side and I have some ideas I wanted to test out but often the models I work with need to maintain quad topology.

Would tri-mesh be suitable for manipulating quad meshes & potentially ngons or does it implement some of the groundwork in a way that makes this impractical?

Thanks

**Comment by asny** [comment, marked as answer]
Hi @ant1fact 

Great to hear that you found it interesting. Can I ask why you need to maintain a quad mesh? I don't think it is feasible to change `tri-mesh` to work directly on quad meshes. You could add a conversion function to and from quad mesh but that would introduce some error and it is not trivial to go from a triangle mesh to a quad mesh.

**Comment by ant1fact** [comment]
Hi @asny 

First of all thank you for getting back & taking the time to be curious why quads are even a thing :)

Quad meshes are very common in movies, games, and motion graphics but also used a lot in architectural design. Some of the advantages of quad meshes are (but not limited to):

- Better deformation of animated meshes
- Optimal subdivision using Catmull-Clark / OpenSubdiv
- Better shading, especially for highly reflective surfaces like chrome
- Most of 3D production is done using quads in software like Maya, 3ds Max, Houdini, Blender, Modo, etc. (Triangulation happens either at export or only at render time by the GPU)
- Quad meshes are much easier to edit and reuse

If you allow, I will post an example by a well-known 3d artist in the film industry. These are some 3d parts that were built using the same techniques that are used in film production:
https://www.artstation.com/marketplace/p/00x/sub-d-film-production-kit

Based on my limited research and understanding, the half-edge data structure seems to be the most performant of all the common 3d data structures so it would make sense if some of the aforementioned 3d applications were also using it. Hence my idea that it should then also be possible to extend the functionality to quads and beyond. (Please excuse my obviously flawed reasoning, but I hope you get me:)

Kind regards,
David

**Reply by asny** [reply]
Hi David.

No problem, I am always curious to know how people use tools in this area. I certainly know that quad meshes is a thing, but didn't know all of the advantages, so thanks for that info. But what I wanted to ask you (which wasn't really clear) was what you need to do, what operations on a quad mesh do you need?

Anyway, most of the operations rely on the mesh consisting of triangles so you need to re-implement most of them for quad meshes and then you can just as well create a new crate (`quad-mesh`?). Specifically edge flip is used a lot which does not have a quad mesh equivalent. I guess you could use a half-edge data structure but that is not the biggest part of `tri-mesh` so maybe just copy that part to a new crate.

Best regards, Asger.

**Reply by ant1fact** [reply]
Hi Asger,

sorry, I could have been a bit more specific. What I hope to do - as a hobby a project & to better learn rust - is try to create some operations on 3d meshes which aren't usually available in the most commonly used 3d modelling software.

One example would be easily capping a cylinder with a quads-only half sphere. Usually this process involves many steps.

In any case, I really appreciate the suggestion to take the basic structure and re-implement functionality on top of it. This is the kind of information I was looking for as this will save me from taking a wrong path and losing unnecessary time. I had the feeling that by default a crate called "tri-mesh" may not be the best option for quad+ meshes but I also hoped that at least some of it might prove useful:)

Best wishes,
David

**Reply by asny** [reply]
No problem at all. I think it makes sense to see if something can be reused, but sometimes there is not enough overlap and I think this is one of the cases. I actually thought about it when naming the crate, so yeah, I would copy the basic data structure if it also makes sense in practice and reimplement the functionality 👍 Anyway, sounds like a fun project 🥳 

- Asger.
